lib/rubocop/cop/style/blocks.rb

```diff
@@ -74,6 +89,97 @@ def parentheses?(send_node)
         def operator?(method_name)
           method_name =~ /^\W/
         end
+
+        def proper_block_style?(node)
+          case style
+          when :weirich
+            weirich_block_style?(node)
+          when :multiline
+            multiline_block_style?(node)
+          end
+        end
+
+        def weirich_block_style?(node)
+          method_name = extract_method_name_from_block(node)
+          return true if ignored_method?(method_name)
+
+          block_begin = node.loc.begin.source
+
+          if block_begin == '{'
+            functional_method?(method_name) || functional_block?(node)
+          else
+            procedural_method?(method_name) || !return_value_used?(node)
+          end
+        end
+
+        def multiline_block_style?(node)
+          block_length = Util.block_length(node)
+          block_begin = node.loc.begin.source
+
+          if block_length > 0
+            block_begin != '{'
+          else
+            block_begin == '{'
+          end
+        end
+
+        def extract_method_name_from_block(block)
+          node, _args, _body = *block
+          _receiver, method_name, *_args = *node
+
+          method_name
+        end
+
+        def ignored_method?(method_name)
+          ignored_methods.include?(method_name)
+        end
+
+        def functional_method?(method_name)
+          functional_methods.include?(method_name)
+        end
+
+        def functional_block?(node)
+          return_value_used?(node) || return_value_of_scope?(node)
+        end
+
+        def procedural_method?(method_name)
+          procedural_methods.include?(method_name)
+        end
+
+        def return_value_used?(node)
+          return unless node.parent
+
+          # If there are parentheses around the block, check if that
+          # is being used.
+          if node.parent.begin_type?
+            return_value_used?(node.parent)
+          else
+            Util::ASGN_NODES.include?(node.parent.type) ||
+              node.parent.send_type?
+          end
+        end
+
+        def return_value_of_scope?(node)
+          return unless node.parent
+
+          conditional?(node.parent) || node.parent.children.last == node
+        end
+
+        def procedural_methods
+          Array(cop_config['ProceduralMethods']).map(&:to_sym)
```

Comment:
It's a little unnecessary to convert the list to an array. I think we can expect it to be an array. And specs don't cover the case when it's not.